components/layout/display_list_builder.rs

```diff
@@ -1894,83 +1920,134 @@ pub trait BlockFlowDisplayListBuilding {
-                if child.context_type == StackingContextType::PseudoFloat {
-                    new_context.children.push(child);
-                } else {
-                    non_floating_children.push(child);
-                }
+        let scroll_root_id = ScrollRootId::new_of_type(self.fragment.node.id() as usize,
+                                                       self.fragment.fragment_type());
+        let coordinate_system = if self.fragment.establishes_stacking_context() {
+            CoordinateSystem::Own
+        } else {
+            CoordinateSystem::Parent
+        };
+
+        let border_box = self.fragment.stacking_relative_border_box(
+            &self.base.stacking_relative_position,
+            &self.base.early_absolute_position_info.relative_containing_block_size,
+            self.base.early_absolute_position_info.relative_containing_block_mode,
+            coordinate_system);
+
+        let parent_scroll_root_id = state.current_scroll_root_id;
+        state.add_scroll_root(
+            ScrollRoot {
+                id: scroll_root_id,
+                parent_id: parent_scroll_root_id,
+                clip: self.fragment.stacking_relative_content_box(&border_box),
+                size: self.base.overflow.scroll.size,
             }
+        );
+        scroll_root_id
+    }
 
-            parent.add_child(new_context);
-            parent.children.append(&mut non_floating_children);
-            return;
+    fn create_pseudo_stacking_context_for_block(&mut self,
+                                                stacking_context_id: StackingContextId,
+                                                parent_stacking_context_id: StackingContextId,
+                                                parent_scroll_root_id: ScrollRootId,
+                                                state: &mut DisplayListBuildState) {
+        let creation_mode = if self.base.flags.contains(IS_ABSOLUTELY_POSITIONED) ||
+                               self.fragment.style.get_box().position != position::T::static_ {
+            StackingContextCreationMode::PseudoPositioned
+        } else {
+            assert!(self.base.flags.is_float());
+            StackingContextCreationMode::PseudoFloat
+        };
+
+        let new_context = self.fragment.create_stacking_context(stacking_context_id,
+                                                                &self.base,
+                                                                ScrollPolicy::Scrollable,
+                                                                creation_mode,
+                                                                parent_scroll_root_id);
+        state.add_stacking_context(parent_stacking_context_id, new_context);
+
+        self.base.collect_stacking_contexts_for_children(state);
+
+        let new_children: Vec<StackingContext> =
+            state.stacking_context_children(stacking_context_id).drain(..).collect();
```

Comment:
ditto here. Also, if you're emptying the children list, you might as well use `remove(stacking_context_id)` to avoid using `collect`, right?

You could write something like:

```rust
fn take_stacking_context_children_for(&mut self, id: StackingContextId) -> Vec<StackingContext> {
    self.stacking_context_children.remove(id).unwrap_or_else(Vec::new)
}
```